core/base/inc/ROOT/TExecutor.hxx

```diff
@@ -8,187 +8,316 @@
-/// \brief This class defines an interface to execute the same task
-/// multiple times in parallel, possibly with different arguments every
-/// time. The classes implementing it mimic the behaviour of python's pool.Map method. 
-///
-/// ###ROOT::TExecutor::Map
-/// The two possible usages of the Map method are:\n
-/// * Map(F func, unsigned nTimes): func is executed nTimes with no arguments
-/// * Map(F func, T& args): func is executed on each element of the collection of arguments args
-///
-/// For either signature, func is executed as many times as needed by a pool of
-/// nThreads threads; It defaults to the number of cores.\n
-/// A collection containing the result of each execution is returned.\n
-/// **Note:** the user is responsible for the deletion of any object that might
-/// be created upon execution of func, returned objects included: ROOT::TExecutor never
-/// deletes what it returns, it simply forgets it.\n
-///
-/// \param func
-/// \parblock
-/// a lambda expression, an std::function, a loaded macro, a
-/// functor class or a function that takes zero arguments (for the first signature)
-/// or one (for the second signature).
-/// \endparblock
-/// \param args
-/// \parblock
-/// a standard vector, a ROOT::TSeq of integer type or an initializer list for the second signature.
-/// An integer only for the first.\n
-/// \endparblock
-///
-/// **Note:** in cases where the function to be executed takes more than
-/// zero/one argument but all are fixed except zero/one, the function can be wrapped
-/// in a lambda or via std::bind to give it the right signature.\n
-///
-/// #### Return value:
-/// An std::vector. The elements in the container
-/// will be the objects returned by func.
-
-namespace ROOT {
-
-template<class subc>
-class TExecutor {
+
+namespace ROOT{
+
+namespace Internal{
+class TExecutor: public TExecutorBaseImpl<TExecutor> {
 public:
-   explicit TExecutor() = default;
-   explicit TExecutor(size_t /* nThreads */ ){};
 
-   template< class F, class... T>
-   using noReferenceCond = typename std::enable_if<"Function can't return a reference" && !(std::is_reference<typename std::result_of<F(T...)>::type>::value)>::type;
+   explicit TExecutor(unsigned nProcessingUnits = 0) :
+      TExecutor(ROOT::IsImplicitMTEnabled() ? ROOT::Internal::ExecutionPolicy::kMultithread : ROOT::Internal::ExecutionPolicy::kSerial, nProcessingUnits) {}
+
+   explicit TExecutor(ROOT::Internal::ExecutionPolicy execPolicy, unsigned nProcessingUnits = 0) : fExecPolicy(execPolicy) {
+      fExecPolicy = execPolicy;
+      switch(fExecPolicy) {
+         case ROOT::Internal::ExecutionPolicy::kSerial:
+            fSeqPool = std::unique_ptr<ROOT::TSequentialExecutor>(new ROOT::TSequentialExecutor());
+            break;
+#ifdef R__USE_IMT
+         case ROOT::Internal::ExecutionPolicy::kMultithread:
+            fThreadPool = std::unique_ptr<ROOT::TThreadExecutor>(new ROOT::TThreadExecutor(nProcessingUnits));
+            break;
+#endif
```

Comment:
What happens when IMT is off, and somebody passes `kMT`? The switch doesn't have a default case.

Reply:
It does 🤔 

Is it possible that you are commenting on an outdated commit?